packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart

```diff
@@ -0,0 +1,1296 @@
+  //
+  // This value is only reset when the timer between a [PointerUpEvent] and the
+  // [PointerDownEvent] times out or when a new [PointerDownEvent] is tracked in
+  // [addAllowedPointer].
+  PointerUpEvent? get currentUp => _up;
+
+  // The number of consecutive taps that the most recently tracked [PointerDownEvent]
+  // in [currentDown] represents.
+  //
+  // This value defaults to zero, meaning a tap series is not currently being tracked.
+  //
+  // When this value is greater than zero it means [addAllowedPointer] has run
+  // and at least one [PointerDownEvent] belongs to the current series of taps
+  // being tracked.
+  //
+  // [addAllowedPointer] will either increment this value by `1` or set the value to `1`
+  // depending if the new [PointerDownEvent] is determined to be in the same series as the
+  // tap that preceded it. If too much time has elapsed between two taps, the recognizer has lost
+  // in the arena, the gesture has been cancelled, or the recognizer is being disposed then
+  // this value will be set to `0`, and a new series will begin.
+  int get consecutiveTapCount => _consecutiveTapCount;
+
+  // The set of [LogicalKeyboardKey]s pressed when the most recent [PointerDownEvent]
+  // was tracked in [addAllowedPointer].
+  //
+  // This value defaults to an empty set.
+  //
+  // When the timer between two taps elapses, the recognizer loses the arena, the gesture is cancelled
+  // or the recognizer is disposed of then this value is reset.
+  Set<LogicalKeyboardKey> get keysPressedOnDown => _keysPressedOnDown ?? <LogicalKeyboardKey>{};
+
+  // The upper limit for the [consecutiveTapCount]. When this limit is reached
+  // all tap related state is reset and a new tap series is tracked.
+  //
+  // If this value is null, [consecutiveTapCount] can grow infinitely large.
+  int? get maxConsecutiveTap;
+
+  // The maximum distance in logical pixels the gesture is allowed to drift
+  // from the initial touch down position before the [consecutiveTapCount]
+  // and [keysPressedOnDown] are frozen and the remaining tracker state is
+  // reset. These values remain frozen until the next [PointerDownEvent] is
+  // tracked in [addAllowedPointer].
+  double? get slopTolerance;
+
+  // Private tap state tracked.
+  PointerDownEvent? _down;
+  PointerUpEvent? _up;
+  int _consecutiveTapCount = 0;
+  Set<LogicalKeyboardKey>? _keysPressedOnDown;
+
+  bool _wonArena = false;
+  OffsetPair? _originPosition;
+  int? _previousButtons;
+
+  // For timing taps.
+  Timer? _consecutiveTapTimer;
+  Offset? _lastTapOffset;
+
+  // When tracking a tap, the [consecutiveTapCount] is incremented if the given tap
+  // falls under the tolerance specifications and reset to 1 if not.
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    super.addAllowedPointer(event);
+    if (maxConsecutiveTap == _consecutiveTapCount) {
+      _tapTrackerReset();
+    }
+    _up = null;
+    _wonArena = false;
```

Comment:
It is possible that `acceptGesture` is called without a `PointerUpEvent` being received. This can happen if the arena is sweeped before a `PointerUpEvent` is routed, and this recognizer is the only one in the arena.

Reply:
I added a test to verify this.

Reply:
I am for changing the name to `_isPointerDown` or `_sentTapDown` but am not sure about the `_isPointerDown = !(_up != null && _down != null);`.

Reply:
> _isPointerDown = !(_up != null && _down != null);

What I meant is the name also matches the logic in acceptGesture

`_isPointerDown = !(_up != null && _down != null);` is logically equivalent to

```dart
_isPointerDown = true;
if (_up != null && _down != null) {
 _isPointerDown= false;
}
```
assuming it is impossible that _up == null and _down ==null